src/Simulator/FluidSimulator.cpp

```diff
@@ -128,100 +128,94 @@ bool cFluidSimulator::IsHigherMeta(NIBBLETYPE a_Meta1, NIBBLETYPE a_Meta2)
+	if (!IsAllowedBlock(BlockID))  // No Fluid -> No Flowing direction :D
 	{
-		return Y_MINUS;
+		return vDirection;
 	}
-	*/
 
-	NIBBLETYPE LowestPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z);  // Current Block Meta so only lower points will be counted
-	int X = 0, Z = 0;  // Lowest Pos will be stored here
+	NIBBLETYPE CentralPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z);
 
-	if (IsAllowedBlock(m_World.GetBlock(a_X, a_Y + 1, a_Z)) && a_Over)  // check for upper block to flow because this also affects the flowing direction
+	if (CentralPoint > 7)
 	{
-		return GetFlowingDirection(a_X, a_Y + 1, a_Z, false);
+		CentralPoint -= 8;
 	}
 
-	std::vector< Vector3i * > Points;
-
-	Points.reserve(4);  // Already allocate 4 places :D
+	// blocks around the checking pos
 
-	// add blocks around the checking pos
-	Points.push_back(new Vector3i(a_X - 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X + 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z + 1));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z - 1));
+	Points[0] = { a_X + 1, a_Y, a_Z };
+	Points[1] = { a_X, a_Y, a_Z + 1 };
+	Points[2] = { a_X - 1, a_Y, a_Z };
+	Points[3] = { a_X, a_Y, a_Z - 1 };
 
-	for (auto itr = Points.cbegin(), end = Points.cend(); itr != end; ++itr)
+	for (int i = 0; i < 4; i++)
 	{
-		Vector3i * Pos = (*itr);
-		auto PosBlockID = m_World.GetBlock(Pos->x, Pos->y, Pos->z);
-		if (IsAllowedBlock(PosBlockID))
+		BlockID = m_World.GetBlock(Points[i].x, Points[i].y, Points[i].z);
+
+		if (IsAllowedBlock(BlockID))
 		{
-			NIBBLETYPE Meta = m_World.GetBlockMeta(Pos->x, Pos->y, Pos->z);
+			LevelPoint[i] = m_World.GetBlockMeta(Points[i].x, Points[i].y, Points[i].z);
 
-			if (Meta > LowestPoint)
+			if (LevelPoint[i] > 7)
+			{
+				LevelPoint[i] -= 8;
+			}
+			else if (LevelPoint[i] == 0)
 			{
-				LowestPoint = Meta;
-				X = Pos->x;
-				Z = Pos->z;
+				BlockID = m_World.GetBlock(Points[i].x, Points[i].y - 1, Points[i].z);
+
+				if (IsAllowedBlock(BlockID))
+				{
+					LevelPoint[i] = CentralPoint + 1;
+				}
 			}
 		}
-		else if (PosBlockID == E_BLOCK_AIR)
+		else if (BlockID == E_BLOCK_AIR)
 		{
-			LowestPoint = 9;  // This always dominates
-			X = Pos->x;
-			Z = Pos->z;
+			BlockID = m_World.GetBlock(Points[i].x, Points[i].y - 1, Points[i].z);
 
+			if (IsAllowedBlock(BlockID))
+			{
+				LevelPoint[i] = CentralPoint + 1;
+			}
+			else
+			{
+				LevelPoint[i] = CentralPoint;
+			}
+		}
+		else
+		{
+			LevelPoint[i] = CentralPoint;
 		}
-		delete Pos;
-		Pos = nullptr;
 	}
 
-	if (LowestPoint == m_World.GetBlockMeta(a_X, a_Y, a_Z))
-	{
-		return NONE;
-	}
+	// Calculate the flow direction
 
-	if (a_X - X > 0)
-	{
-		return X_MINUS;
-	}
+	vDirection.x = LevelPoint[0] - LevelPoint[2];
+	vDirection.z = LevelPoint[1] - LevelPoint[3];
 
-	if (a_X - X < 0)
-	{
-		return X_PLUS;
-	}
+	float Length = static_cast<float>(sqrt(vDirection.x * vDirection.x + vDirection.z * vDirection.z));
 
-	if (a_Z - Z > 0)
+	if (Length != 0.0f)
```

Comment:
```cpp
if (Direction.HasNonZeroLength())
{

Reply:
This code is faster (only two operands, not three).

Reply:
The root isn't calculated twice because you don't need the exact length for `HasZeroLength`.
As for the number of operands, you'd maybe save at best one or two CPU cycles. I don't think it's worth it.